File: Data_Structure/week1_basic_data_structures/1_brackets_in_code/check_brackets.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;
struct Bracket {
    Bracket(char type, int position):
        type(type),
        position(position)
    {}

    bool Matchc(char c) {
        if (type == '[' && c == ']')
            return true;
        if (type == '{' && c == '}')
            return true;
        if (type == '(' && c == ')')
            return true;
        return false;
    }

    char type;
    int position;
};

int main() {
    std::string text;
    getline(std::cin, text);
    int answer = -1;
    std::stack <Bracket> opening_brackets_stack;
    for (int position = 0; position < text.length(); ++position) {
        char next = text[position];

        if (next == '(' || next == '[' || next == '{') {
            // Process opening bracket, write your code here
            Bracket *tmp = new Bracket(next, position+1);
            opening_brackets_stack.push(*tmp);
        }

        if (next == ')' || next == ']' || next == '}') {
            // Process closing bracket, write your code here
            if(! opening_brackets_stack.empty()){    
                if(!opening_brackets_stack.top().Matchc(next)){
                    answer = position + 1;
                    break;
                }
                opening_brackets_stack.pop();
            }
            else{
                answer = position + 1;
                break;
        }
    }
    }
    if(opening_brackets_stack.empty() && answer == -1){
        cout << "Success" << endl;
    }
    else{
        if(opening_brackets_stack.size() && answer == -1){
            answer = opening_brackets_stack.top().position;
        }
        cout << answer << endl;
    }
    // Printing answer, write your code here

    return 0;
}

File: Data_Structure/week1_basic_data_structures/1_brackets_in_code/check_brackets.py
# python3

from collections import namedtuple


Bracket = namedtuple("Bracket", ["char", "position"])
answer = -1
len_stack = 0
def are_matching(left, right):
    return (left + right) in ["()", "[]", "{}"]


def find_mismatch(text):
    opening_brackets_stack = []
    global answer
    global len_stack
    for i, next in enumerate(text):
        if next in "([{":
            # Process opening bracket, write your code here
            opening_brackets_stack.append(Bracket(next,i))
            pass

        if next in ")]}":
            # Process closing bracket, write your code here
            if len(opening_brackets_stack)==0:
                answer = i + 1
                break
            top = opening_brackets_stack.pop().char
            if not are_matching(top, next):
                answer = i + 1
                break 
            pass
    len_stack = len(opening_brackets_stack)
    if len_stack:
        answer = opening_brackets_stack.pop().position + 1

def main():
    text = input()
    mismatch = find_mismatch(text)
    global answer
    # Printing answer, write your code here
    if answer != -1:
        print(answer)
    else:
        print("Success")

if __name__ == "__main__":
    main()
